bokehjs/src/lib/core/visuals.ts

```diff
@@ -8,6 +8,166 @@ import {LineJoin, LineCap, FontStyle, TextAlign, TextBaseline} from "./enums"
 import {HasProps} from "./has_props"
 import {ColumnarDataSource} from "models/sources/columnar_data_source"
 
+function _horz(ctx: CanvasRenderingContext2D, h: number, h2: number): void {
+  ctx.moveTo(0, h2+0.5)
+  ctx.lineTo(h, h2+0.5)
+  ctx.stroke()
+}
+
+function _vert(ctx: CanvasRenderingContext2D, h: number, h2: number): void {
+  ctx.moveTo(h2+0.5, 0)
+  ctx.lineTo(h2+0.5, h)
+  ctx.stroke()
+}
+
+function _x(ctx: CanvasRenderingContext2D, h: number): void {
+  ctx.moveTo(0, h)
+  ctx.lineTo(h, 0)
+  ctx.stroke()
+  ctx.moveTo(0, 0)
+  ctx.lineTo(h, h)
+  ctx.stroke()
+}
+
+function _get_canvas(size: number): HTMLCanvasElement {
+  const canvas = document.createElement('canvas')
+  canvas.width = size
+  canvas.height = size
+  return canvas
+}
+
+function create_hatch_canvas(hatch_pattern: string, hatch_color: string, hatch_scale: number, hatch_weight:number ): HTMLCanvasElement {
```

Comment:
`hatch_pattern` should be of type `HatchPatternType`. `hatch_color` should be of `Color` type (event if that directly aliases to `string`). Note that at use site you will have to use `!`, to reduce `Color | null` to `Color` (this is a side effect of having `doit` and retrieving color value independently).